experiments/process_sandbox/cxxapi/cxxsandbox.h

```diff
@@ -0,0 +1,304 @@
+     */
+    Ret (*function)(Args...);
+    /**
+     * The type encoding string for this function.
+     */
+    char* type_encoding() override
+    {
+      return strdup(typeid(function).name());
+    }
+    /**
+     * Call the function and, if it returns a non-void type, store the result
+     * into the frame.
+     */
+    void operator()(void* callframe) override
+    {
+      assert(callframe != nullptr);
+      auto frame = (static_cast<CallFrame*>(callframe));
+      if constexpr (!std::is_void_v<Ret>)
+      {
+        frame->ret = std::apply(function, frame->args);
+      }
+      else
+      {
+        std::apply(function, frame->args);
+      }
+    }
+
+  public:
+    /**
+     * Construct a new sandbox-exported function.  This should be called only by
+     * the `export_function` in `ExportedLibrary`, but that calls it indirectly
+     * and so we can't sensibly do this via a friend definition.
+     */
+    SandboxExportedFunction(Ret (*fn)(Args...)) : function(fn) {}
+  };
+
+  /**
+   * Class wrapping a set of functions that are exported from a library.  This
+   * class should never be instantiated directly.  A library that is intended
+   * to be used in a sandboxed context should implement a function with the
+   * following name and signature:
+   *
+   * ```
+   * extern "C" void sandbox_init(sandbox::ExportedLibrary* library);
+   * ```
+   *
+   * The sandbox runner process will call this function, providing it with a
+   * pointer to an instance of this class, during setup.
+   */
+  class ExportedLibrary
+  {
+    /**
+     * The vtable for this library: map from integers to functions that can be
+     * invoked.
+     */
+  public:
+    inline static std::vector<std::unique_ptr<SandboxExportedFunctionBase>>
+      functions;
+
+    static void call(int idx, void* args)
+    {
+      (*functions[idx])(args);
+    }
+    /**
+     * Returns the type encoding string (in a C++ ABI-specific format) of the
+     * function at the specified vtable index.
+     */
+    static char* type_encoding(int idx)
+    {
+      return functions.at(idx)->type_encoding();
+    }
+    /**
+     * Export a function to consumers of this sandboxed library.  The function
+     * is inserted in the next vtable entry so the order of calls to this
+     * method from inside the sandbox must match the order of
+     * `SandboxedFunction` objects created for a `SandboxedLibrary` that
+     * corresponds to the outside of this sandbox.
+     */
+    template<typename Ret, typename... Args>
+    static void export_function(Ret (*fn)(Args...))
+    {
+      if (
+        (functions.size() == 0) &&
+        (fn != reinterpret_cast<Ret (*)(Args...)>(type_encoding)))
+      {
+        export_function(type_encoding);
+      }
+      functions.emplace_back(new SandboxExportedFunction(fn));
+    }
+  };
+
+  extern "C" void sandbox_call(int idx, void* args)
+  {
+    ExportedLibrary::call(idx, args);
```

Comment:
It's unclear to me what `ExportedLibrary` is.

Being a static class means all users of `sandbox_call` will use the same exported library interface.

I expected that this code would be the base for each actual library (a Verona module in the source path) and each separate module would have their own dispatchers.

Do you expect the Verona compiler to generate this for every module at compile time? If not, how are we going to separate each module's implementation? Namespaces?

Reply:
When you say "a single sandbox_call ... from the library", do you mean "one dispatcher per foreign module, so all sandbox regions using the same module will re-use the same dispatcher"?

I think this is the same issue I'm having with the other comment on the `main` function, and I think here you have answered that question: no, this isn't how we'll do in the compiler. But the scope of the dispatcher is still confusing to me.

I guess, as an experiment, this makes sense and works, so I don't want to hold on to this. We can discuss how we'll implement in the compiler at a later time, no worries.